include/gunrock/algorithms/hits.hxx

```diff
@@ -0,0 +1,375 @@
+    }
+    os<<"===Hub\n\n";
+    for(int i = 0; i < this->hub.size(); i++){
+      os<<"vertex ID: "<<this->hub_vertex[i]<<std::endl;
+      os<<"authority: "<<this->hub[i]<<std::endl;
+    }
+  }
+
+  void print_result(int max_vertices, std::ostream& os = std::cout){
+    int vertices = (max_vertices < this->hub.size())
+              ? max_vertices
+              : this->hub.size();
+    os<<"===Authority\n\n";
+    for(int i = 0; i < vertices; i++){
+      os<<"vertex ID: "<<this->auth_vertex[i]<<std::endl;
+      os<<"authority: "<<this->auth[i]<<std::endl;
+    }
+    os<<"===Hub\n\n";
+    for(int i = 0; i < vertices; i++){
+      os<<"vertex ID: "<<this->hub_vertex[i]<<std::endl;
+      os<<"authority: "<<this->hub[i]<<std::endl;
+    }
+  }
+
+
+  // For internal use
+  thrust::device_vector<float>
+  get_auth(){
+  return this->auth;
+  }
+
+  thrust::device_vector<float>
+  get_hub(){
+  return this->hub;
+  }
+};// end of result_c
+
+template <typename graph_t>
+struct problem_t: gunrock::problem_t<graph_t> {
+public:
+  using vertex_t = typename graph_t::vertex_type;
+  using edge_t = typename graph_t::edge_type;
+  using weight_t = typename graph_t::weight_type;
+
+  int iterator = 0;
+  int max_iterations;
+
+  // Sum of squares of authority and hub
+  // qqq find a way to reduce mem access
+  float sum = 0;
+
+  vertex_t vertex_num = 0;
+
+  thrust::device_vector<float> auth_curr;
+  thrust::device_vector<float> hub_curr;
+  thrust::device_vector<float> auth_next;
+  thrust::device_vector<float> hub_next;
+
+  // poniters to the data inside the device_vector
+  float* auth_curr_p = nullptr;
+  float* hub_curr_p = nullptr;
+  float* auth_next_p = nullptr;
+  float* hub_next_p = nullptr;
+
+  // Functor for deviding each element with sum and taking the sqrt
+  template<typename T>
+  struct op{
+  private:
+    float sum;
+  public:
+    op(float sum) : sum(sum){}
+    __device__ __host__
+    T operator()(const T& x) const {
+      return sqrt(x/this->sum);
+      }
+  };
+
+  problem_t(graph_t& G,
+            std::shared_ptr<cuda::multi_context_t> _context,
+            int max_iterations)
+      : gunrock::problem_t<graph_t>(G, _context),
+        max_iterations(max_iterations){
+  vertex_num = G.get_number_of_vertices();
+  auto policy = _context->get_context(0)->execution_policy();
+
+  auth_curr.resize(vertex_num);
+  thrust::fill(policy, auth_curr.begin(), auth_curr.end(), 0);
+  auth_next.resize(vertex_num);
+  thrust::fill(policy, auth_next.begin(), auth_next.end(), 0);
+  hub_curr.resize(vertex_num);
+  thrust::fill(policy, hub_curr.begin(), hub_curr.end(), 0);
+  hub_next.resize(vertex_num);
+  thrust::fill(policy, hub_next.begin(), hub_next.end(), 0);
+  }
+
+  void init() override{
+    auth_curr_p = auth_curr.data().get();
+    hub_curr_p = hub_curr.data().get();
+    auth_next_p = auth_next.data().get();
+    hub_next_p = hub_next.data().get();
+  }
+  void reset() override{}
```

Comment:
The initial fills from https://github.com/gunrock/essentials/pull/83/files#diff-b62c127579eb994567fe8e548f95a87eba8fc88110834e01736092e7d816e04fR166-R172 can be part of `reset` instead of being part of the constructor. `reset()` should run on every run of the application to "reset" the intermediate data I am assuming?